### PR TITLE
Multi Networks Support

### PR DESCRIPTION
**Pull request by AlonaKaplan**
The purpose of this PR is to extend ovn-kubernetes to support multiple networks.
The secondary networks will be added using multus (or any other solution complies with the de-facto standard).

It means the pods annotation should contain the required network attachments under - 'k8s.v1.cni.cncf.io/networks'.

```apiVersion: v1
kind: Pod
metadata:
  name: pod-with-multiple-ovn-networks
  annotations:
    k8s.v1.cni.cncf.io/networks: '[
            { "name": "blue-network-attachment”},
            { "name": "red-network-attachment”}
    ]'
   switches: [“blue-network”, “red-network”]
...
```

The network attachment CRD should contain the name of the logical switch represents the logical network in the config section.

```apiVersion: "k8s.cni.cncf.io/v1"
kind: NetworkAttachmentDefinition
metadata:
  name: “blue-network-attachment”
spec: 
  config: '{
    “name”: “blue-network”
    "cniVersion": "0.3.1",
    "type": "ovn-k8s-cni-overlay",
    "Ipam":{},
    "dns":{}
  }'
```

Once the pod will be started, multus will invoke on the minion the binary 'ovn-k8s-cni-overlay' per each pod's network attachment with the type 'ovn-k8s-cni-overlay'.

The ovn-kubernetes code is changed in the following way -

Master

1. **Once the 'watcher' gets a 'pod ADD' event** from k8s API server, it will read the 'switches' annotation from the pod and **create a logical port per switch** in OVN. **For a logical switch named “logicalSwitchName”, the name of the logical port will be "namespace_podName_logicalSwitchName"**. The default logical port will be created the same way as before (the port will be called "namespace_podName" and will be created on the logical switch having the name of the node).
Each newly created logical port in OVN will have "dynamic" address (OVN will assign per each port with “dynamic” address an IP address and MAC address).
3. **Per each port, the watcher will write the assigned IP/MAC into the pod object annotation under the key "ovn_exrta".**

```apiVersion: v1
kind: Pod
metadata:
  name: pod-with-multiple-ovn-networks
  annotations:
    k8s.v1.cni.cncf.io/networks: '[
            { "name": "blue-network-attachment"},
            { "name": "red-network-attachment"},
              ]'
    ...

   switches: [“blue-network”, “red-network”]

   ovn = 
    {"ip_address":"10.244.0.5/24",
     "mac_address":"0a:00:00:00:00:15",
     "gateway_ip": "10.244.0.1"}

   ovn_extra=
    {"blue-network" :
      {"ip_address":"192.168.0.2/24",
       "mac_address":"0a:00:00:00:00:01",
       "gateway_ip": ""}
      },
    "red-network":
      {"ip_address":"10.234.0.3/24",
       "mac_address":"0a:00:00:00:00:02",
       "gateway_ip": ""
      },
      …
```

Minion
1. **Multus invokes the 'ovn-k8s-cni-overlay' per each interface** (of course only network interfaces that are using network attachment with the type - 'ovn-k8s-cni-overlay').
2. **The CNI binary ('ovn-k8s-cni-overlay') reads the annotation of the relevant network from the pod** object 'ovn_extra' section and sets up the network namespace of the container.
3. A veth pair is created in the container and then the host end is moved into host netns.
4. **A new port is created on “br-int”** connecting the host-end to the switch **with "external_ids:iface-id: namespace_podName_logicalSwitchName”** (the logicalSwitchName is taken from the NetworkAttachment and passed by multus to the minion binary (same as the other values from the config section of the network attachment).

Notes -
1. The secondary ovn-kubernetes networks can be added whether ovn-kubernetes is the primary network or not.
However, currently, the master side will create all the 'default' network infrastructure on the ovn north db even if ovn-kubernetes is not used as the primary network. The minion side won't be invoked for the default network in case ovn-kubernetes is not the primary cni.
2. Interfaces on the same pod connected to the same network/logical switch are currently not supported (since the name of logical port is namespace_podName_switchName and we cannot have multiple ports with the same name).
2. Reviewers, please share your thoughts regarding - 
   2.1 Instead of the 'switches' annotation on the pod, taking the switch name directly from the NetworkAttachment CRD.
  2.2 Instead of the 'ovn_extra' annotation, writing the mac/ip/gateway to the pod's multus annotation.

[5 minutes demo](https://asciinema.org/a/P6r9q59sxQxrY5wgfdnYbAFnw
)
[For more details](https://docs.google.com/document/d/1M2SFLEvp34GPZnhEf_D8ZeMY3ZAztgiulN5J-SKHj3Q/edit?ts=5bacc332#)

### CODE REVIEW

File: go-controller/pkg/cni/cni.go [hunk 1]
```diff
@@ -183,4 +212,4 @@ func HandleCNIRequest(request *PodRequest) ([]byte, error) {
 	}
 	logrus.Infof("Returning pod network request %v, result %s err %v", request, string(result.Response), result.Err)
 	return result.Response, result.Err
-}
+}
```
Comment (phoracek) [inline go-controller/pkg/cni/cni.go]: Please add new line

File: go-controller/pkg/cni/cni.go [hunk 2]
```diff
@@ -93,24 +103,43 @@ func (pr *PodRequest) cmdAdd() *PodResult {
 		return nil
 	}
 
-	ovnAnnotation, ok := annotation["ovn"]
-	if !ok {
-		logrus.Errorf("failed to get ovn annotation from pod")
-		return nil
-	}
+	var ovnAnnotation string
```
Comment (phoracek) [inline go-controller/pkg/cni/cni.go]: You don't need to declare it here, it is used only in the body of the first if block.

File: go-controller/pkg/cni/cni.go [hunk 3]
```diff
@@ -93,24 +103,43 @@ func (pr *PodRequest) cmdAdd() *PodResult {
...
+	var ok bool
```
Comment (phoracek) [inline go-controller/pkg/cni/cni.go]: Same as the previous comment

File: go-controller/pkg/cni/cni.go [hunk 4]
```diff
@@ -93,24 +103,43 @@ func (pr *PodRequest) cmdAdd() *PodResult {
...
+		}
 
-	var ovnAnnotatedMap map[string]string
-	err = json.Unmarshal([]byte(ovnAnnotation), &ovnAnnotatedMap)
-	if err != nil {
-		logrus.Errorf("unmarshal ovn annotation failed")
-		return nil
+		err = json.Unmarshal([]byte(ovnAnnotation), &ovnNetworkAnnotatedMap)
+		if err != nil {
+			logrus.Errorf("unmarshal ovn annotation failed")
```
Comment (phoracek) [inline go-controller/pkg/cni/cni.go]: You should also log the error message:

logrus.Errorf("unmarshal ovn annotation failed: %v", err)

File: go-controller/pkg/cni/cni.go [hunk 5]
```diff
@@ -93,24 +103,43 @@ func (pr *PodRequest) cmdAdd() *PodResult {
...
+		ovnExtraAnnotation, ok := annotation[ovnAnnotationKey]
+		if !ok {
+			logrus.Errorf("failed to get ovn_extra annotation from pod")
+			return nil
+		}
+		var ovnExtraAnnotatedMap map[string]map[string]string
+
+		err = json.Unmarshal([]byte(ovnExtraAnnotation), &ovnExtraAnnotatedMap)
+		if err != nil {
+			logrus.Errorf("unmarshal ovn_extra annotation failed = %v", err)
```
Comment (phoracek) [inline go-controller/pkg/cni/cni.go]: use ': ' instead of ' = '

File: go-controller/pkg/cni/helper_linux.go [hunk 6]
```diff
@@ -123,6 +139,7 @@ func (pr *PodRequest) ConfigureInterface(namespace string, podName string, macAd
 		fmt.Sprintf("external_ids:ip_address=%s", ipAddress),
 		fmt.Sprintf("external_ids:sandbox=%s", pr.SandboxID),
 	}
+
```
Comment (phoracek) [inline go-controller/pkg/cni/helper_linux.go]: unrelated change

File: go-controller/pkg/cni/helper_linux.go [hunk 7]
```diff
@@ -150,16 +168,32 @@ func (pr *PodRequest) ConfigureInterface(namespace string, podName string, macAd
...
-	ifaceName := pr.SandboxID[:15]
-	ovsArgs := []string{
-		"del-port", "br-int", ifaceName,
+	networkName := pr.CNIConf.Name
+	var portList []string
+	if networkName == "ovn-kubernetes" {
+		portList = make([]string, 1)
+		portList[0] = pr.SandboxID[:15]
+	} else {
+		var err error
```
Comment (phoracek) [inline go-controller/pkg/cni/helper_linux.go:177]: you can just use := ovsFind, can't you?

File: go-controller/pkg/cni/helper_linux.go [hunk 8]
```diff
@@ -150,16 +168,32 @@ func (pr *PodRequest) ConfigureInterface(namespace string, podName string, macAd
 
 // PlatformSpecificCleanup deletes the OVS port
 func (pr *PodRequest) PlatformSpecificCleanup() error {
-	ifaceName := pr.SandboxID[:15]
-	ovsArgs := []string{
-		"del-port", "br-int", ifaceName,
+	networkName := pr.CNIConf.Name
+	var portList []string
+	if networkName == "ovn-kubernetes" {
```
Comment (phoracek) [inline go-controller/pkg/cni/helper_linux.go]: please use the same helper as before isDefaultInterface, or better, create a helper function for it, since it's used in multiple places

File: go-controller/pkg/ovn/common.go [hunk 9]
```diff
@@ -220,20 +212,58 @@ func (oc *Controller) getIPFromOvnAnnotation(ovnAnnotation string) string {
...
+}
 
-	return ovnAnnotationMap["mac_address"]
+func (oc *Controller) getNetworkInfoFromOvnAnnotation(ovnAnnotation string, networkName string) map[string]string {
+	if ovnAnnotation == "" {
+		return nil
+	}
+
+	var ovnExtraAnnotationMap map[string]map[string]string
+	err := json.Unmarshal([]byte(ovnAnnotation), &ovnExtraAnnotationMap)
```
Comment (phoracek) [inline go-controller/pkg/ovn/common.go:251]: You can do:

```
if err := json.Unmarshal([]byte(ovnAnnotation), &ovnExtraAnnotationMap); err != nil {
    log, return error
}

normal return
```

File: go-controller/pkg/ovn/common.go [hunk 10]
```diff
@@ -220,20 +212,58 @@ func (oc *Controller) getIPFromOvnAnnotation(ovnAnnotation string) string {
...
+	if err != nil {
+		logrus.Errorf("Error in json unmarshaling ovn annotation  (%v)", err)
+		return nil
+	}
+
+	return ovnExtraAnnotationMap[networkName]
+}
+
+func (oc *Controller) getNetworkNamesFromPodAnnotations(podAnnotations map[string]string) []string {
+	switches_annotation, hasExtraSwitches := podAnnotations["switches"]
```
Comment (phoracek) [inline go-controller/pkg/ovn/common.go]: switchesAnnotation

File: go-controller/pkg/ovn/ovn.go [hunk 11]
```diff
@@ -116,20 +118,56 @@ func (oc *Controller) Run() error {
 	return nil
 }
 
+func (oc *Controller) addDefaultLogicalPort(pod *kapi.Pod) {
+	annotation := oc.addLogicalPort(pod, pod.Spec.NodeName)
+	_, isStaticIP := pod.Annotations["ovn"]
```
Comment (phoracek) [inline go-controller/pkg/ovn/ovn.go:127]: as previously, you can use one-line if

File: go-controller/pkg/ovn/common.go [hunk 12]
```diff
@@ -220,20 +212,58 @@ func (oc *Controller) getIPFromOvnAnnotation(ovnAnnotation string) string {
...
+	if hasExtraSwitches {
```
Comment (phoracek) [inline go-controller/pkg/ovn/common.go:263]: you can keep it on one line with assignment

File: go-controller/pkg/ovn/ovn.go [hunk 13]
```diff
@@ -116,20 +118,56 @@ func (oc *Controller) Run() error {
...
+	switches := oc.getNetworkNamesFromPodAnnotations(pod.Annotations)
+	if switches != nil {
+		annotations := make(map[string]string)
+		for _, logicalSwitch := range switches {
+			annotations[logicalSwitch] = oc.addLogicalPort(pod, logicalSwitch)
+		}
+
+		ovnExtraAnnotation := fmt.Sprintf("{")
+		for network, networkAnnotation := range annotations {
+			ovnExtraAnnotation = fmt.Sprintf("%s\\\"%s\\\":%s, ", ovnExtraAnnotation, network, networkAnnotation)
```
Comment (phoracek) [inline go-controller/pkg/ovn/ovn.go:149]: don't you need to put between quotes also the value?

File: go-controller/pkg/ovn/ovn.go [hunk 14]
```diff
@@ -116,20 +118,56 @@ func (oc *Controller) Run() error {
...
+		}
+
+		ovnExtraAnnotation = strings.TrimSuffix(ovnExtraAnnotation, ", ")
+		ovnExtraAnnotation = fmt.Sprintf("%s}", ovnExtraAnnotation)
+
+		err := oc.kube.SetAnnotationOnPod(pod, "ovn_extra", ovnExtraAnnotation)
+		if err != nil {
+			logrus.Errorf("Failed to set annotation on pod %s - %v", pod.Name, err)
```
Comment (phoracek) [inline go-controller/pkg/ovn/ovn.go]: please put error after `: `

File: go-controller/pkg/ovn/pods.go [hunk 15]
```diff
@@ -11,16 +13,25 @@ import (
...
-	// get the list of logical switch ports (equivalent to pods)
+	// get the list of logical switch ports
 	expectedLogicalPorts := make(map[string]bool)
 	for _, podInterface := range pods {
 		pod, ok := podInterface.(*kapi.Pod)
 		if !ok {
 			logrus.Errorf("Spurious object in syncPods: %v", podInterface)
 			continue
 		}
+
```
Comment (phoracek) [inline go-controller/pkg/ovn/pods.go]: unrelated change

**Review by phoracek — CHANGES_REQUESTED** [review]
Only partial review

File: go-controller/pkg/cni/cni.go [hunk 16]
```diff
@@ -61,6 +61,10 @@ func extractPodBandwidthResources(podAnnotations map[string]string) (int64, int6
 func (pr *PodRequest) cmdAdd() *PodResult {
 	namespace := pr.PodNamespace
 	podName := pr.PodName
+
+	networkName := pr.CNIConf.Name
+	logrus.Infof("Adding interface for network = %s", networkName)
+
 	if namespace == "" || podName == "" {
```
Comment (SchSeba) [inline go-controller/pkg/cni/cni.go]: Need to add the `networkName  == "" ` here

File: go-controller/pkg/cni/cni.go (hunk 5 above)
Comment (SchSeba) [reply go-controller/pkg/cni/cni.go]: maybe change the error to `failed to unmarshal ovn_extra annotationfrom pod`

File: go-controller/pkg/cni/helper_linux.go [hunk 17]
```diff
@@ -101,20 +104,33 @@ func setupInterface(netns ns.NetNS, containerID, ifName, macAddress, ipAddress,
...
+// TODO alona update the window_helper
+func (pr *PodRequest) ConfigureInterface(namespace string, podName string, networkName string, macAddress string, ipAddress string, gatewayIP string, mtu int, ingress, egress int64) ([]*current.Interface, error) {
 	netns, err := ns.GetNS(pr.Netns)
 	if err != nil {
 		return nil, fmt.Errorf("failed to open netns %q: %v", pr.Netns, err)
 	}
 	defer netns.Close()
 
-	hostIface, contIface, err := setupInterface(netns, pr.SandboxID, pr.IfName, macAddress, ipAddress, gatewayIP, mtu)
+	isDefaultInterface := networkName == "ovn-kubernetes"
```
Comment (SchSeba) [inline go-controller/pkg/cni/helper_linux.go]: create a const variable for `ovn-kubernetes`

File: go-controller/pkg/cni/helper_linux.go [hunk 18]
```diff
@@ -150,16 +168,32 @@ func (pr *PodRequest) ConfigureInterface(namespace string, podName string, macAd
...
-		"del-port", "br-int", ifaceName,
+	networkName := pr.CNIConf.Name
+	var portList []string
+	if networkName == "ovn-kubernetes" {
+		portList = make([]string, 1)
+		portList[0] = pr.SandboxID[:15]
+	} else {
+		var err error
+		ifaceName := fmt.Sprintf("%s_%s_%s", pr.PodNamespace, pr.PodName, networkName)
+		portList, err = ovsFind("interface", "name", "external-ids:iface-id="+ifaceName)
```
Comment (SchSeba) [inline go-controller/pkg/cni/helper_linux.go]: can you please use 
```golang 
fmt.Sprintf("external-ids:iface-id=%s",ifaceName)
```

File: go-controller/pkg/cni/helper_linux.go [hunk 19]
```diff
@@ -150,16 +168,32 @@ func (pr *PodRequest) ConfigureInterface(namespace string, podName string, macAd
...
+		if err != nil {
+			logrus.Infof("failed to find OVS port with external-ids:iface-id:=%s  %v", ifaceName, err)
```
Comment (SchSeba) [inline go-controller/pkg/cni/helper_linux.go]: just add `error:` before the error string
```golang
logrus.Infof("failed to find OVS port with external-ids:iface-id:=%s error: %v", ifaceName, err)
```

File: go-controller/pkg/cni/cni.go [hunk 20]
```diff
@@ -93,24 +103,43 @@ func (pr *PodRequest) cmdAdd() *PodResult {
...
+		}
+		var ovnExtraAnnotatedMap map[string]map[string]string
+
+		err = json.Unmarshal([]byte(ovnExtraAnnotation), &ovnExtraAnnotatedMap)
+		if err != nil {
+			logrus.Errorf("unmarshal ovn_extra annotation failed = %v", err)
+			return nil
+		}
+
+		ovnNetworkAnnotatedMap, ok = ovnExtraAnnotatedMap[networkName]
```
Comment (yuvalif) [inline go-controller/pkg/cni/cni.go:133]: "ok" is unused, please use _ instead

File: go-controller/pkg/cni/helper_linux.go [hunk 21]
```diff
@@ -92,7 +95,7 @@ func setupInterface(netns ns.NetNS, containerID, ifName, macAddress, ipAddress,
 	}
 
 	// rename the host end of veth pair
-	hostIface.Name = containerID[:15]
+	hostIface.Name = hostIfaceName
```
Comment (yuvalif) [inline go-controller/pkg/cni/helper_linux.go:98]: why not truncate at 15 chars?

File: go-controller/pkg/cni/cni.go [hunk 22]
```diff
@@ -73,6 +77,12 @@ func (pr *PodRequest) cmdAdd() *PodResult {
 	}
 	kubecli := &kube.Kube{KClient: clientset}
 
+	isDefaultInterface := networkName == "ovn-kubernetes"
```
Comment (yuvalif) [inline go-controller/pkg/cni/cni.go:80]: this "is default" check is done in two places. probably better to pout in a helper function

Comment (yuvalif) [reply go-controller/pkg/cni/cni.go:80]: see: https://github.com/AlonaKaplan/ovn-kubernetes/pull/3/files#diff-0f28a8316299be5031f7bf5ec31b441aR115

File: go-controller/pkg/ovn/pods.go [hunk 23]
```diff
@@ -149,64 +183,99 @@ func (oc *Controller) deleteLogicalPort(pod *kapi.Pod) {
...
+		}
+		oc.deletePodFromNamespaceAddressSet(pod.Namespace, ipAddress)
 	}
-	oc.deletePodFromNamespaceAddressSet(pod.Namespace, ipAddress)
+
 	return
 }
 
-func (oc *Controller) addLogicalPort(pod *kapi.Pod) {
+func (oc *Controller) addLogicalPort(pod *kapi.Pod, logicalSwitch string) string {
```
Comment (yuvalif) [inline go-controller/pkg/ovn/pods.go]: IMO, better to use newAnotation as a named return value

**Review by SchSeba — COMMENTED** [review]
Partial review only

File: go-controller/pkg/ovn/pods.go [hunk 24]
```diff
@@ -149,64 +183,99 @@ func (oc *Controller) deleteLogicalPort(pod *kapi.Pod) {
...
+		portName = fmt.Sprintf("%s_%s_%s", pod.Namespace, pod.Name, logicalSwitch)
+	}
+
 	logrus.Debugf("Creating logical port for %s on switch %s", portName, logicalSwitch)
 
-	annotation, isStaticIP := pod.Annotations["ovn"]
+	var ovnNetworkAnnotatedMap map[string]string
+	var isStaticIP bool
+	if isDefaultPort {
+		var annotation string
```
Comment (yuvalif) [inline go-controller/pkg/ovn/pods.go:265]: no need to define the var here, just use := in the next line

File: go-controller/pkg/ovn/pods.go [hunk 25]
```diff
@@ -231,16 +300,25 @@ func (oc *Controller) addLogicalPort(pod *kapi.Pod) {
...
 				"stdout: %q, stderr: %q (%v)",
 				portName, out, stderr, err)
-			return
+			return ""
 		}
 	}
 
 	oc.logicalPortCache[portName] = logicalSwitch
 
 	gatewayIP, mask, err := oc.getGatewayFromSwitch(logicalSwitch)
```
Comment (yuvalif) [inline go-controller/pkg/ovn/pods.go:309]: do you need to run getGateway for the non default port?

File: go-controller/pkg/cni/helper_linux.go [hunk 26]
```diff
@@ -101,20 +104,33 @@ func setupInterface(netns ns.NetNS, containerID, ifName, macAddress, ipAddress,
...
 	}
 	defer netns.Close()
 
-	hostIface, contIface, err := setupInterface(netns, pr.SandboxID, pr.IfName, macAddress, ipAddress, gatewayIP, mtu)
+	isDefaultInterface := networkName == "ovn-kubernetes"
+	var ifaceID string
+	var hostIfaceName string
+	if isDefaultInterface {
+		ifaceID = fmt.Sprintf("%s_%s", namespace, podName)
+		hostIfaceName = pr.SandboxID[:15]
```
Comment (yuvalif) [inline go-controller/pkg/cni/helper_linux.go:119]: this is likely to exceed 15 chars. IMO, we should not silently truncate as different pods may step on one another. some options:
 - if the name is longer than 15 chars, submit an error (not so good, as it will require changes to pod names and/or namespaces)
 - use a hash function instead of truncation, makes collisions very unlikely (problem is that names are becoming meaningless)

File: go-controller/pkg/cni/cni.go (hunk 1 above)
Comment (AlonaKaplan) [reply go-controller/pkg/cni/cni.go]: done

File: go-controller/pkg/cni/cni.go (hunk 2 above)
Comment (AlonaKaplan) [reply go-controller/pkg/cni/cni.go]: done

File: go-controller/pkg/cni/cni.go (hunk 3 above)
Comment (AlonaKaplan) [reply go-controller/pkg/cni/cni.go]: done

File: go-controller/pkg/cni/cni.go (hunk 4 above)
Comment (AlonaKaplan) [reply go-controller/pkg/cni/cni.go]: done

File: go-controller/pkg/cni/cni.go (hunk 5 above)
Comment (AlonaKaplan) [reply go-controller/pkg/cni/cni.go]: done

File: go-controller/pkg/cni/helper_linux.go (hunk 6 above)
Comment (AlonaKaplan) [reply go-controller/pkg/cni/helper_linux.go]: done

File: go-controller/pkg/cni/helper_linux.go (hunk 7 above)
Comment (AlonaKaplan) [reply go-controller/pkg/cni/helper_linux.go:177]: I cannot since the 'portList' is needed outside the scope of the 'if'.
portList, err = ovsFind("interface", "name", fmt.Sprintf("external-ids:iface-id=%s",ifaceName))

File: go-controller/pkg/cni/helper_linux.go (hunk 8 above)
Comment (AlonaKaplan) [reply go-controller/pkg/cni/helper_linux.go]: done

File: go-controller/pkg/ovn/common.go (hunk 10 above)
Comment (AlonaKaplan) [reply go-controller/pkg/ovn/common.go]: done

File: go-controller/pkg/ovn/ovn.go (hunk 13 above)
Comment (AlonaKaplan) [reply go-controller/pkg/ovn/ovn.go:149]: The value is not a string, it is a map.

File: go-controller/pkg/ovn/ovn.go (hunk 14 above)
Comment (AlonaKaplan) [reply go-controller/pkg/ovn/ovn.go]: done

File: go-controller/pkg/cni/cni.go (hunk 20 above)
Comment (AlonaKaplan) [reply go-controller/pkg/cni/cni.go:133]: now it is used:)

File: go-controller/pkg/cni/helper_linux.go (hunk 21 above)
Comment (AlonaKaplan) [reply go-controller/pkg/cni/helper_linux.go:98]: It was moved to ConfigureInterface.

File: go-controller/pkg/cni/cni.go (hunk 22 above)
Comment (AlonaKaplan) [reply go-controller/pkg/cni/cni.go:80]: done

File: go-controller/pkg/ovn/pods.go (hunk 23 above)
Comment (AlonaKaplan) [reply go-controller/pkg/ovn/pods.go]: done

File: go-controller/pkg/cni/cni.go (hunk 16 above)
Comment (AlonaKaplan) [reply go-controller/pkg/cni/cni.go]: done

File: go-controller/pkg/cni/helper_linux.go (hunk 17 above)
Comment (AlonaKaplan) [reply go-controller/pkg/cni/helper_linux.go]: Moved to a helper method.

File: go-controller/pkg/cni/helper_linux.go (hunk 18 above)
Comment (AlonaKaplan) [reply go-controller/pkg/cni/helper_linux.go]: done

File: go-controller/pkg/cni/helper_linux.go (hunk 19 above)
Comment (AlonaKaplan) [reply go-controller/pkg/cni/helper_linux.go]: done

File: go-controller/pkg/ovn/pods.go (hunk 24 above)
Comment (AlonaKaplan) [reply go-controller/pkg/ovn/pods.go:265]: The next line is annotation, isStaticIP = pod.Annotations["ovn"]
and isStaticIP is needed outside of the 'if' scope.

File: go-controller/pkg/ovn/pods.go (hunk 25 above)
Comment (AlonaKaplan) [reply go-controller/pkg/ovn/pods.go:309]: Yes, it may have gateway. But unlike the default one, it may also have no gateway.

File: go-controller/pkg/cni/helper_linux.go (hunk 26 above)
Comment (AlonaKaplan) [reply go-controller/pkg/cni/helper_linux.go:119]: The sandbox id is not a name, it is more like a uuid.
But anyway, this is the old code for the default pod network. I don't want to change it in the scope of this PR.

**Review by yuvalif — APPROVED** [review]
